### DISCUSSION TITLE
iTwin migration from 2.x to 3.x

### DISCUSSION THREAD

**Discussion by ec-dev-1** [iTwin/itwinjs-core#4482] [Q&A] [answered]
Hello everyone,

We are currently in the process of upgrading our iTwin to 3.x
I was able to follow the instructions with using the codemod tool.
I also manually changed anything else that was giving errors that the codemod tool did not catch.

I'm able to start the app with no errors until the Bentley authentication.
Although it seems that providing a valid username/pw goes through, we appear to receive errors regarding fetching the iTwin model.

`Error opening the iModel connection: Unauthorized: IDX10214: Audience validation failed. Audiences: 'https://ims.bentley.com/as/token.oauth2, https://imsoidc.bentley.com/as/token.oauth2, https://imsoidc.bentley.com/resources, itwinjs-services'. Did not match: validationParameters.ValidAudience: 'null' or validationParameters.ValidAudiences: 'bentley-api-management'.`

The IMJS_AUTH_CLIENT_REDIRECT_URI is  {{my dev port}} /signin-callback

Within the public folder, the signin-callback folder contains the same files from the iTwin repo which are:
-index.html
-oidc-client.min.js

I may have missed it but were we supposed to update anything within the signin-callback folder?

Any help would be much appreciated.

Thank you,

![Bentley-error](https://user-images.githubusercontent.com/86491141/195411029-14b4e0ae-207f-40b2-a058-1a381c4d0e9f.png)

**Comment by aruniverse** [comment, marked as answer]
Can you go to https://developer.bentley.com/apps/spa/{YOUR_CLIENT_ID_HERE}, and check your client doesn't have a banner that says something like "Upgrade" ? Can you also make sure you have the valid scopes for 3.x?

**Reply by aruniverse** [reply]
You will need `realitydata:read imodels:read imodelaccess:read` but you probably only have `itwinjs` (which can be removed)

**Reply by ec-dev-1** [reply]
There's no reference to "Upgrade" when I visit my App Spa's page.
It also looks like I already have those scope added.

Here's a complete list of what I currently have:
`imodelaccess:read realitydata:read realitydata:modify imodels:modify email projects:read profile openid users:read imodels:read projects:modify itwinjs organization`

Thanks for looking into this post. 

Looking forward to solving this issue!

**Reply by aruniverse** [reply]
Can you remove: `email profile openid itwinjs organization`

**Reply by aruniverse** [reply]
What is your issuer? `ims.bentley.com` or `imsoidc.bentley.com` ?

**Reply by ec-dev-1** [reply]
Ok, I will update the scope shortly but here's the answer to your other question:

IMJS_AUTH_AUTHORITY="https://ims.bentley.com"

update:
Thought I'd add this in.  Does this look correct? :
` BrowserAuthorizationCallbackHandler.handleSigninCallback(redirectURI).then(
    () => {
      client.signIn();
    }
  );`

**Reply by ec-dev-1** [reply]
@aruniverse  Wanted to quickly update you that I made some progress.  Although the scopes were correct on the My Apps section of the Developer page, I neglected to copy these to the IMJS_AUTH_CLIENT_SCOPES of my .env.

The error regarding authentication is now gone. 
I'm now faced with a JS error regarding the PlanarClipMask but I think I can handle it from here.


Thank you once again for helping me find my way around the iTwin Platform.

**Comment by JiriJa** [comment]
Hi @ec-dev-1  @aruniverse
Please can I ask You?    I´m currently in the process of upgrading my iTwin to 3.x and you have mentioned that You have been able to follow the instructions with using the codemod tool. Please can You share this instructions or link to them? Many thanks Jiri.

Note:
I´m going to change my old notebook for newest one. Do you have any step-by-step guidance how to get successes with this process? All my attempts ends in wrong way ☹ 

Now on my new PC I have installed from scratch VCS + node.js + iTwin.js 3.4.1 and then I have fetched my code from repository but my local host:3000 iTwin viewer is always with white blank rendering.
On my old PC I have been installed all needed since 2021 VCS + node. js + iTwin.js 2.?.? … and this installation is from June 2021 and still is working well, but only on old machine.

Best Regards
Jiri.

**Reply by aruniverse** [reply]
Have you looked at the dev tools console? Are you seeing any messages in there?

**Comment by JiriJa** [comment]
Still is someting out of order. 
@aruniverse, so I have finelly decided to create new react app from scratch according to the guidace from an Accreditaton Developer "iTwin Platform" course  .... so after "npx create-react-app smart-house-itwin-viewer --template @itwin/web-viewer --scripts-version @bentley/react-scripts" I have set my "Authorization Client Settings" - token sequece number, Test ids : IMJS_CONTEXT_ID, IMJS_IMODEL_ID and then I have been runing  "npm start" command from new Terminal .... so ... result is ... Terminal responded "Starting the development server..." with respond to the "http://localhost:3000/"... then I heave signed in by my Bentley User account ... Terminal reply: "Compiled successfully! You can now view smart-house-itwin-viewer in the browser. ... But  localhost:3000 has been rendered by white sreen only. Any idea please?
![image](https://user-images.githubusercontent.com/34089657/199274917-dd17b0a2-d18f-4bc0-b39e-97593cdd53e7.png)

**Reply by aruniverse** [reply]
What does the dev tools console in your browser look like? 
This is what i'm seeing when I bootstrap a new app without filling in the .env at all.
![image](https://user-images.githubusercontent.com/11051042/199278100-fe73d2e8-7d78-418a-b693-a4494b762e89.png)

Take a look at your [index.tsx](https://github.com/iTwin/viewer/blob/master/packages/modules/cra-template-web-viewer/template/src/index.tsx) and make sure the following env vars are set.

Regarding the _"process is not defined"_ error, thats a known issue, take a look at the second bullet point [here](https://github.com/iTwin/viewer/blob/master/releases/CHANGELOG-2.0.md#known-issues).

**Reply by JiriJa** [reply]
@aruniverse 
 ... my console looks like this:
![image](https://user-images.githubusercontent.com/34089657/199281228-2831d200-0389-40be-9475-7f25b5a58353.png)

**Reply by aruniverse** [reply]
have you tried adding the **IMJS_ITWIN_ID** env var like the error says? 
https://github.com/iTwin/viewer/blob/master/packages/modules/cra-template-web-viewer/template/src/App.tsx#L74

**Reply by JiriJa** [reply]
@aruniverse  ... please, where am I able to find correct IMJS_ITWIN_ID number? .... because on my old PC ( app on iTwin 2.x???) occurs olny this "IDS" variables IMJS_CONTEXT_ID= ....  and IMJS_IMODEL_ID= ... 
![image](https://user-images.githubusercontent.com/34089657/199284897-03d46905-d55c-453b-b622-cf6747ee2dbf.png)

**Reply by aruniverse** [reply]
We renamed _Context_ to _iTwin_ in iTwin.js 3.0. https://github.com/iTwin/viewer/blob/master/releases/CHANGELOG-2.0.md#renamed
So just renamed _IMJS_CONTEXT_ID_ to _IMJS_ITWIN_ID_

These are the ids you'd get if you were to copy from the My iModels page or the [Create iModel](https://developer.bentley.com/tutorials/web-application-quick-start/#4-create-an-imodel) step in developer.bentley.com, 

<img width="410" alt="image" src="https://user-images.githubusercontent.com/11051042/199286360-ebc16f41-62c8-4cf0-9a2b-78f99c7e237f.png">

**Reply by JiriJa** [reply]
Yes, IMJS_CONTEXT_ID to IMJS_ITWIN_ID works well ... but next trouble is with error: "The user is unauthorized. Please provide valid authentication credentials." ... I have checked that code is with  authClient prop is not it true? What do You mean?

console:
![image](https://user-images.githubusercontent.com/34089657/199290367-b8c644f4-9f56-4ee8-933e-0b82852491fe.png)

code in App.tsx: 
![image](https://user-images.githubusercontent.com/34089657/199291481-b177d89b-68ab-4424-92a6-80978acd6746.png)

**Reply by aruniverse** [reply]
Please take a look at the [Changelog](https://github.com/iTwin/viewer/blob/master/releases/CHANGELOG-2.0.md#new-scopes-required) for upgrading to iTwin.js 3.0 in Viewer 2.0. Your scopes are incorrect according to the earlier screenshot you shared.

**Reply by JiriJa** [reply]
ok, i´m going to have a look ...  so can I simply delete whole directory with this "react-app smart-house-itwin-viewer" or is there any sophisticated procedure how to do it with somethig like this? : ""npx DELETE-react-app smart-house-itwin-viewer" ??? @aruniverse, Many thanks!

**Reply by aruniverse** [reply]
why do you want to delete the entire directory/app ? just update your scopes by editing your existing app's client registration or [create a new one here](https://developer.bentley.com/tutorials/web-application-quick-start/#3-register-an-application).

**Comment by JiriJa** [comment]
"react-app smart-house-itwin-viewer" I have created only for  temporary purposes "how to do it" ... I hope I  will abel to find correct way how to upgrade my other app "react-app iot-openroads-czech" to the iTwin 3.x  ....

**Comment by JiriJa** [comment]
@aruniverse
I have made some little progress ... but, please can I ask You how can I get missing  **dependencies** like { **Select**, **Toggle** } from modules "@bentley/ui-core" and import them correctly to the project? Same situation with import { RenderMode } from "@bentley/imodeljs-common"; and from "@bentley/imodeljs-frontend" and from "@bentley/ui-abstract"

Cannot find module '@bentley/ui-core' or its corresponding type declarations.
Cannot find module '@bentley/imodeljs-frontend' or its corresponding type declarations
Cannot find module '@bentley/imodeljs-frontend' or its corresponding type declarations.

**Reply by aruniverse** [reply]
As mentioned by Colin, our packages were renamed for 3.0; please take a look at the [changelog](https://github.com/iTwin/itwinjs-core/blob/master/docs/changehistory/3.0.0.md#package-name-changes).

There is a [codemod tool](https://github.com/iTwin/codemods) to help assist you in upgrading to 3.0 and auto update/rename all the packages.

**Reply by JiriJa** [reply]
@aruniverse  @ColinKerr  OK, please, is there any YouTube video how this sophisticated [codemod tool](https://github.com/iTwin/codemods) simply use? ... I´m not sure that this short [README.md](https://github.com/iTwin/codemods#readme) is enough comfortable guidance  for me and my skills and knowledge .... Jiri

**Reply by pmconne** [reply]
Did you try it? What was the result? If you had problems, describe them. If you didn't understand a part of the instructions, tell us which part(s).

**Reply by aruniverse** [reply]
No, unfortunately there are no videos on how it works. The readme should be enough; what issues are you having with it?
Just need to clone the repo, install the dependencies, and then run the script with the required args

**Reply by JiriJa** [reply]
@aruniverse @pmconne 
I have done this with tihs results:
 
I have done "Clone this repo (@itwin/codemods) into the same **parent** directory as the repo I wish to transform" 
result .... now exist ...  C:\Users\jiri.jarolim\codemods   AND  ... C:\Users\jiri.jarolim\iot-openroads-czech     and they    ARE    on the  same directory level as my repo

inside VSC in folder "C:\Users\jiri.jarolim\codemods" I have opened new Terminal and lunch this command:
 **npm install --production**

In the root of my project's repo (C:\Users\jiri.jarolim\iot-openroads-czech), I have run codemods with this found paths:
I have copied together this paths:
C:\Users\jiri.jarolim\iot-openroads-czech\package.json
C:\Users\jiri.jarolim\iot-openroads-czech\tsconfig.json
C:\Users\jiri.jarolim\iot-openroads-czech\src

... inside VSC in folder (C:\Users\jiri.jarolim\iot-openroads-czech) I have opened new Terminal and lunch this single command:
node ../codemods/run-codemods.js C:\Users\jiri.jarolim\iot-openroads-czech\package.json C:\Users\jiri.jarolim\iot-openroads-czech\tsconfig.json C:\Users\jiri.jarolim\iot-openroads-czech\src

with this result:
![image](https://user-images.githubusercontent.com/34089657/200580986-04e74f4c-546c-474f-b387-98a275433e50.png)

Maybe mistake by colne procedure  .... First I have created new directory first and than run git clone ... due to this path exist with double "codemodes":
C:\Users\jiri.jarolim\**codemods**\codemods\run-codemods.js
So, I have modified single command to this and run again:
node ../ codemods/codemods/run-codemods.js C:\Users\jiri.jarolim\iot-openroads-czech\package.json C:\Users\jiri.jarolim\iot-openroads-czech\tsconfig.json C:\Users\jiri.jarolim\iot-openroads-czech\src

with this result:
![image](https://user-images.githubusercontent.com/34089657/200583083-5f409aef-71ed-421a-805e-cb8ffdf7f7f1.png)


still is something wrong ... :-(

**Reply by JiriJa** [reply]
@aruniverse @pmconne 

I never give up ... next attempt ...

new set of command sounds like this:
node ../codemods/codemods/run-codemods.js package.json tsconfig.json src

and Result is now different:
Windows PowerShell
Copyright (C) Microsoft Corporation. All rights reserved.

Try the new cross-platform PowerShell https://aka.ms/pscore6

PS C:\Users\jiri.jarolim\iot-openroads-czech> node ../codemods/codemods/run-codemods.js package.json tsconfig.json src
Processing 14 files... 
Spawning 14 workers...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
[BABEL] Note: The code generator has deoptimised the styling of C:\Users\jiri.jarolim\codemods\codemods\node_modules\typescript\lib\typescript.js as it exceeds the max of 500KB.
All done. 
Results:
0 errors

**Reply by JiriJa** [reply]
@aruniverse @pmconne 
**an output still continues .... is this correct end of whole procedure ?**

0 errors
7 unmodified
0 skipped
7 ok
Time elapsed: 37.483seconds 
Processing 6 files... 
Spawning 6 workers...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
All done. 
Results: 
0 errors
6 unmodified
0 skipped
0 ok
Time elapsed: 1.777seconds
Processing 8 files... 
Spawning 8 workers...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Sending 1 files to free worker...
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
Browserslist: caniuse-lite is outdated. Please run:
  npx browserslist@latest --update-db
  Why you should do it regularly: https://github.com/browserslist/browserslist#browsers-data-updating
All done. 
Results: 
0 errors
4 unmodified
0 skipped
4 ok
Time elapsed: 1.489seconds
Updating package.json
PS C:\Users\jiri.jarolim\iot-openroads-czech>

**Reply by aruniverse** [reply]
The first time you ran it successfully completed:
```
All done.
Results:
0 errors
```
The second time you ran it there were no changes to make since you already it 
```
All done.
Results:
0 errors
4 unmodified
0 skipped
4 ok
```

**Comment by ColinKerr** [comment]
A number of packages were renamed between 2.x and 3.x. You can find all of the renames here: https://www.itwinjs.org/changehistory/3.0.0/#package-name-changes


I think this list covers all of the ones you listed above:
@bentley/ui-core -> @itwin/core-react
@bentley/imodeljs-frontend -> @itwin/core-frontend
@bentley/imodeljs-common -> @itwin/core-common
@bentley/ui-abstract -> @itwin/appui-abstract

**Reply by JiriJa** [reply]
@aruniverse @pmconne

- 'Environment' is not exported from '@itwin/core-frontend'
- Select, Toggle are NOT included in "@itwin/core-react"?

![image](https://user-images.githubusercontent.com/34089657/200593843-be184cb4-f0da-4676-924c-e1840d5ff748.png)

![image](https://user-images.githubusercontent.com/34089657/200594559-b4c19a24-05a9-4ccd-a994-106f2c12b8ef.png)

**Reply by pmconne** [reply]
Environment moved to core-common.

**Reply by JiriJa** [reply]
![image](https://user-images.githubusercontent.com/34089657/200596372-b3afb9bb-faee-4d34-af06-6f2eaac73c94.png)

**Reply by aruniverse** [reply]
Re the AuthorizationClient changes, I would recommend copying the template code from the updated viewer-templates due to the [auth changes made in iTwin.js@3.0](https://github.com/iTwin/itwinjs-core/blob/master/docs/changehistory/3.0.0.md#authorization-clients) and [auth changes made in iTwin Viewer@2.0](https://github.com/iTwin/viewer/blob/master/releases/CHANGELOG-2.0.md#authorization)

**Reply by JiriJa** [reply]
@aruniverse 
many many changes on AutorizationClient I have done ... so ok, first viewport I have rendered .... but next new kind of errors suddenly occures:
![image](https://user-images.githubusercontent.com/34089657/201074116-72be1e79-2a4f-44e3-8239-54523b040047.png)

 next new kind of errors suddenly occures:
now : Cannot assign to 'acsTriad', 'backgroundMap', 'grid',  ,,,, * because it is a read-only property.ts(ts2540 ... no matching results) ... and No quick fixes available
![image](https://user-images.githubusercontent.com/34089657/201072354-1b38ddc1-51d7-467b-8246-cf455844eb59.png)

**Reply by pmconne** [reply]
ViewFlags [became immutable in 3.0](https://www.itwinjs.org/changehistory/3.0.0/#viewflags).

**Reply by JiriJa** [reply]
Hi @pmconne 
any idea how to rewrite from this to this?

![image](https://user-images.githubusercontent.com/34089657/201109775-06137375-af24-461c-a57a-204be30d262c.png)

**Reply by pmconne** [reply]
Did you read the section of the documentation to which I directly linked? It offers code examples illustrating several answers to your question.

**Reply by JiriJa** [reply]
@pmconne 
How to "Update viewflags" in this documentation ([became immutable in 3.0](https://www.itwinjs.org/changehistory/3.0.0/#viewflags).) is maybe good hidden in "poetry" for me :-( ...  ... or is not this true? - then I apologize .... I have app from accreditation course based on iTwin 2.X and now sipmly want to upgrade app to the iTwin 3.X ... but some files desapears completly whitout notice.

**Reply by pmconne** [reply]
```
vp.viewFlags = vp.viewFlags.copy({
   acsTriad: attrValues.acs,
  backgroundMap: attrValues.backgroundMap,
  // etc...
});
```

**Reply by aruniverse** [reply]
![image](https://user-images.githubusercontent.com/11051042/201119564-c901f969-98e3-4de6-bfbc-f0c8884b2598.png)

**Reply by JiriJa** [reply]
@pmconne @aruniverse 
ok , "Update viewflags" modified,

couple of next changes was needed and code was Compiled successfully!. BUT new render is quite different - look at this at the bottom ...
![image](https://user-images.githubusercontent.com/34089657/201316599-b3cdcd60-a59a-4673-9eea-22268f32af1e.png)

        // vp.viewFlags.acsTriad = on;
        vp.viewFlags = vp.viewFlags.copy({ acsTriad: true });

     // return displayStyle.environment.sky.display;
      return displayStyle.environment.displaySky;
 

      // style.environment = new Environment({ sky: { display: on } });
      style.environment = style.environment.withDisplay({ sky: on });


    // const viewFlags = vp.viewFlags.clone();
    const viewFlags = vp.viewFlags.override({ renderMode: mode });

BUT render is quite different - look at this
iTwin2.X app RENDER looks like:
![image](https://user-images.githubusercontent.com/34089657/201315531-c045576d-d32b-4845-95a9-6ec1570033ee.png)

now after upgrade iTwin3.X app RENDER same data looks like:
![image](https://user-images.githubusercontent.com/34089657/201315951-81f4efc5-6bbb-4209-a8ad-1479c74986a9.png)

original REPO for my iTwin2.X app is [here](https://github.com/JiriJa/IoT-OpenRoads-Czech)

**Reply by pmconne** [reply]
You are not seeing the background map because you must [supply your own API keys](https://www.itwinjs.org/changehistory/3.0.0/#removed-default-api-keys) for things like Bing Maps.
I can't comment on your blue and yellow icons because I don't know where they come from - some decorator of yours?

**Reply by JiriJa** [reply]
@pmconne 
Paul, is there any chance how to stay on my new PC on my old iTwin2.X platform? Sustainable infrastructure depends on Sustainable coding and here it seems that it is really hard to continue with old app on new iTwin3.X platform.

original REPO for my iTwin2.X app is [here](https://github.com/JiriJa/IoT-OpenRoads-Czech)

**Reply by pmconne** [reply]
@JiriJa 3.x contained far too many breaking API changes. We learned our lesson and will not repeat with future major releases. However, we extensively documented those changes and what is required to update.

If you want to stay on 2.x, that's your choice. You will miss out on the past 1.5 years of features, performance improvements, and non-critical bug fixes, as well as those coming in future releases; and when support for 2.x comes to an end you will no longer receive critical bug fixes and security fixes.

**Reply by JiriJa** [reply]
Hi, @pmconne @aruniverse
thank You for Your support and patience with greenhorns :-)

**uffff** ... after next level of deep investigation .... now iTwin migration from 2.x to 3.x is completed

![image](https://user-images.githubusercontent.com/34089657/201686090-c7151b6e-f233-4aaa-a269-7b9f4125d92f.png)
